### PR TITLE
Setup.py -- Argprase, Cmake Args, and caching

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

See #38 and #36 for more context

Rewrite the arg parsing in setup.py to use `argparse` as this allows for easier extension in the future and gives a nice `--help` message!

Adds a command line flag for passing arguments to cmake.  Note that due to `argparse`'s parsing, cmake args must be assed as `--cmake-args="...."`.  Doing `--cmake-args "...."` (no =) won't work :(
This can be passed to pip via the `CMAKE_ARGS` environment variable.

Adds caching of arguments passed to `setup.py` via a simple `json` file written into the build_temp dir.  Arguments can be cached by adding `--cache-args` to the invocation of `setup.py`, any current argument cache is ignored.  I have blacked listed `force_cmake` and `cache_args` from the cache.

@mosra I pointed this PR at your PR as we are both playing with things in setup.py

## How Has This Been Tested

Running around and building things.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -122,32 +174,31 @@ def build_extension(self, ext):
...
-                ["cmake", ext.sourcedir] + cmake_args, cwd=self.build_temp, env=env
+                shlex.split("cmake -H{} -B{}".format(ext.sourcedir, self.build_temp))
+                + cmake_args,
+                env=env,
             )
 
-        subprocess.check_call(
-            ["cmake", "--build", "."] + build_args, cwd=self.build_temp
-        )
+        subprocess.check_call(shlex.split("cmake --build {}".format(self.build_temp)))
```
Comment (mosra) [inline setup.py]: This is missing the `build_args`, so `--config` or `-jN` doesn't get set correctly. This worked for me:

```suggestion
        subprocess.check_call(shlex.split("cmake --build {}".format(self.build_temp)) + build_args)
```

Comment (erikwijmans) [reply setup.py]: Whoops!

**Comment by erikwijmans** [comment]
I’ll think about if there’s an easy way to cache the build_ext options. 

One thing you can do is add those to the setup.cfg as build_ext will read args from there also.

**Comment by erikwijmans** [comment]
Give that a try.  I also made `--build-temp=build` the default as that is nicer for helping people debug things.

**Pull request merged**